### PR TITLE
Add Server Checkout Settings Resource with new plugin framework

### PR DESCRIPTION
**Pull request by guoliangye-okta**
## What

Add the new resource `oktapam_server_checkout_settings` with new plugin framework.

## Jira Ticket
https://oktainc.atlassian.net/browse/OKTA-688669

### CODE REVIEW

File: oktapam/fwprovider/framework_provider.go
Comment (guoliangye-okta) [inline oktapam/fwprovider/framework_provider.go:1]: should we move this to upper module as we talked last time?

File: spec/provider-code-spec.json [hunk 1]
```diff
@@ -32,71 +32,67 @@
...
-															{
-																"path": "github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
-															}
-														],
-														"schema_definition": "stringvalidator.OneOf(\n\"user\",\n\"user_group\",\n\"server\",\n\"project\",\n\"sudo_entitlement\",\n\"resource_group\",\n\"secret_folder\",\n\"secret\",\n)"
-													}
-												}
-											]
-										}
+						"name": "checkout_duration_in_seconds",
```
Comment (guoliangye-okta) [inline spec/provider-code-spec.json:39]: generated from the spec generator, keep it here as an example.

File: spec/openapi.yaml
Comment (guoliangye-okta) [inline spec/openapi.yaml:1]: Update this file with the resource specific to the openAPI endpoint specification from the Platform. Keep it here as an example.

File: spec/generator_config.yaml
Comment (guoliangye-okta) [inline spec/generator_config.yaml:1]: update the generator config with API endpoints will be used for each resource CURD method.
Keep it here for example.

File: spec/Makefile
Comment (guoliangye-okta) [inline spec/Makefile:1]: update resource name here to allow spec generator to generate corresponding scalffold code/spec/schema

File: oktapam/fwprovider/resource_server_checkout_settings.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	}
+
+	// Set state to empty data
+	state.Id = types.StringValue("")
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	// should we delete the state here?
```
Comment (guoliangye-okta) [inline oktapam/fwprovider/resource_server_checkout_settings.go:246]: Should we remove the state of this resource or just update the ID to empty?

File: oktapam/fwprovider/framework_provider.go
Comment (sachinsaxena-okta) [inline oktapam/fwprovider/framework_provider.go:1]: I think we should follow this structure - 

oktapam
      - provider
            - framework
                 - provider.go 
            - sdkv2
                 - provider.go
            - factory.go (This is to create a provider Server Factory. This code currently we are duplicating in provider_test.go and main.go)

Comment (sachinsaxena-okta) [reply oktapam/fwprovider/framework_provider.go:1]: Resources and Resource Test can be at the same place for all the providers. So, basically keep them under - oktapam/<all resources and data sources>. I also observed that you havn't created datasource for this.

File: oktapam/fwprovider/resource_server_checkout_settings.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"checkout_duration_in_seconds": schema.Int64Attribute{
+				Optional:            true,
+				Computed:            true,
+				Description:         "The duration in seconds for the checkout. If the checkout is enabled, the duration is the maximum time a user can access the resource before the checkout expires.",
```
Comment (sachinsaxena-okta) [inline oktapam/fwprovider/resource_server_checkout_settings.go:62]: We are keeping description in oktapam/constants/descriptions

File: oktapam/fwprovider/resource_server_checkout_settings.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+				MarkdownDescription: "If provided, only the account identifiers listed are excluded from the checkout requirement. This list is only considered if `checkout_required` is set to `true`. Only one of `include_list` and `exclude_list` can be specified in a request since they are mutually exclusive.",
+			},
+			"include_list": schema.ListAttribute{
+				ElementType:         types.StringType,
+				Optional:            true,
+				Description:         "If provided, only the account identifiers listed are required to perform a checkout to access the resource. This list is only considered if `checkout_required` is set to `true`. Only one of `include_list` and `exclude_list` can be specified in a request since they are mutually exclusive.",
+				MarkdownDescription: "If provided, only the account identifiers listed are required to perform a checkout to access the resource. This list is only considered if `checkout_required` is set to `true`. Only one of `include_list` and `exclude_list` can be specified in a request since they are mutually exclusive.",
+			},
+			"project": schema.StringAttribute{
```
Comment (sachinsaxena-okta) [inline oktapam/fwprovider/resource_server_checkout_settings.go:85]: This should have force new true. If project is changed in the resource that means deleting the setting for old project and creating new one.

File: oktapam/fwprovider/resource_server_checkout_settings.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+				Required:            true,
+				Description:         "The UUID of a Project",
+				MarkdownDescription: "The UUID of a Project",
+			},
+			"resource_group": schema.StringAttribute{
```
Comment (sachinsaxena-okta) [inline oktapam/fwprovider/resource_server_checkout_settings.go:90]: This should have force new 'true'. If resource group is changed in the resource that means deleting the setting for old project and creating new one.

File: oktapam/fwprovider/resource_server_checkout_settings.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating server checkout settings", err.Error())
+		return
+	}
+	// Set state to fully populated data
+	plan.Id = types.StringValue(formatServerCheckoutSettingsID(plan.ResourceGroup, plan.Project))
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
```
Comment (sachinsaxena-okta) [inline oktapam/fwprovider/resource_server_checkout_settings.go:126]: What is the purpose of doing this? Errors are already added in resp.Diagnostics

File: oktapam/fwprovider/resource_server_checkout_settings.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+	}
+	r.client = provider.SDKClientWrapper
+}
+
+func formatServerCheckoutSettingsID(resourceGroupID string, projectID string) string {
+	// server checkout settings don't have an identifier in itself and is really an attribute of a project.
+	// we manage it as a separate resource since it's lifecycle is somewhat separate from a project.
+	// since project password settings are managed as a separate resource with format resourceGroupID/projectID already,
+	// we can just append the server_checkout_settings to the end of the ID
+	return fmt.Sprintf("%s/%s/server_checkout_settings", resourceGroupID, projectID)
```
Comment (sachinsaxena-okta) [inline oktapam/fwprovider/resource_server_checkout_settings.go:276]: You don't need to add server_checkout_settings in the path.

**Pull request merged**